src/Castle.Windsor.Tests/ReferenceTracker.cs

```diff
@@ -0,0 +1,137 @@
+// limitations under the License.
+
+namespace CastleTests
+{
+	using System;
+
+	using NUnit.Framework;
+
+	public static class ReferenceTracker
+	{
+		/// <summary>
+		/// <para>
+		/// Use a lambda method to evaluate an expression returning an object instance to track.
+		/// </para>
+		/// <para>
+		/// (If the expression is evaluated directly inside the test method, the runtime will keep it alive
+		/// until the test method returns if the runtime is in debug mode. ReSharper’s test runner runs in
+		/// debug mode when the solution is configuration is Debug.)
+		/// </para>
+		/// </summary>
+		public static ReferenceTracker<T> Track<T>(Func<T> getInstanceToTrack) where T : class
+		{
+			return ReferenceTracker<T>.Track(getInstanceToTrack);
+		}
+	}
+
+	/// <summary>
+	/// A test helper encapsulating correct reference tracking.
+	/// </summary>
+	public struct ReferenceTracker<T> where T : class
+	{
+		private readonly WeakReference weakReference;
+
+		private ReferenceTracker(WeakReference weakReference)
+		{
+			this.weakReference = weakReference;
+		}
+
+		/// <summary>
+		/// <para>
+		/// Use a lambda method to evaluate an expression returning an object instance to track.
+		/// </para>
+		/// <para>
+		/// (If the expression is evaluated directly inside the test method, the runtime will keep it alive
+		/// until the test method returns if the runtime is in debug mode. ReSharper’s test runner runs in
+		/// debug mode when the solution is configuration is Debug.)
+		/// </para>
+		/// </summary>
+		public static ReferenceTracker<T> Track(Func<T> getInstanceToTrack)
+		{
+			if (getInstanceToTrack == null)
+				throw new ArgumentNullException(nameof(getInstanceToTrack));
+
+			return new ReferenceTracker<T>(new WeakReference(getInstanceToTrack.Invoke()));
+		}
+
+		/// <summary>
+		/// Calls <see cref="GC.Collect()"/> and asserts that the tracked instance is still alive.
+		/// </summary>
+		public void AssertStillReferenced()
+		{
+			GC.Collect();
+			if (!weakReference.IsAlive)
+				Assert.Fail("The tracked instance is not longer referenced.");
+		}
+
+		/// <summary>
+		/// Calls <see cref="GC.Collect()"/> and asserts that the tracked instance is no longer alive.
+		/// </summary>
+		public void AssertNoLongerReferenced()
+		{
+			GC.Collect();
+			if (weakReference.IsAlive)
+				Assert.Fail("The tracked instance is still referenced.");
+		}
+
+		/// <summary>
+		/// <para>
+		/// Calls <see cref="GC.Collect()"/> and asserts that the tracked instance is still alive and
+		/// passes it to the specified action.
+		/// </para>
+		/// <para>
+		/// Be careful not to let the tracked instance become reachable via a local variable in the test method
+		/// or it will be kept alive until the end of the test method if the runtime is in debug mode.
+		/// (See <see cref="Track"/>.)
+		/// </para>
+		/// </summary>
+		public void AssertStillReferencedAndDo(Action<T> action)
```

Comment:
When I first saw a use of this method, I didn't quite understand why one would need a combination of an `Assert` + a side effect. I guess it's mostly for clean-up. If so, perhaps it would be more self-explanatory to have a separate `CleanUp` method or similar?, even if that results in two lines (instead of one) at the call sites.

Reply:
@stakx This method is essential for the tests which need to act on the reference at a later point in time. The whole point of ReferenceTracker is to avoid evaluating that reference directly in the test method.

As to why it also asserts, it seemed like the obvious thing to do if the reference was no longer kept alive. Other options would be to no-op or to return false, but that just complicates the implementations of the tests.

Yet another alternative which I didn't do because it seemed complex was to extend this helper API to have some sort of keep-alive methods.

Reply:
If it's not trivial to split up `AssertStillReferencedAndDo` into an `AssertStillReferenced` and `CleanUp` (or whatever it would be called), then please ignore my suggestion and leave the method as is. It probably doesn't make any sense to bloat this helper class' API needlessly.

Reply:
Yeah; the specific issue with splitting it into two methods is that it's no longer atomic; there will then be two separate calls to WeakReference.Target and the target may be alive for the assert check and dead for the `Do`/`CleanUp` work. To split it up, we'd need to add KeepAlive-style methods. maybe like:

```c#
var target = ReferenceTracker.Track(() => Container.Resolve<ThingWhichMustNotBeNamed>());

using (target.AssertStillReferencedAndStartKeepingAlive())
{
    target.Do(thing => container.Release(thing));
} // Stop keeping it alive

target.AssertNoLongerReferenced();
```

Oops, now we have `AssertStillReferencedAndStartKeepingAlive` instead of `AssertStillReferencedAndDo`. I didn't think that far ahead, but we can't avoid making a decision about what to do atomically if the thing is no longer alive when the test expects it to be, and asserting seemed like it put the least burden on the test method.